In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
#Imported on 15/02/2024
df = pd.read_csv('https://ai-jobs.net/salaries/download/salaries.csv')
df.head(10)

work_year experience_level employment_type  \
0       2024               EN              FT   
1       2024               EN              FT   
2       2024               SE              FT   
3       2024               SE              FT   
4       2024               SE              FT   
5       2024               SE              FT   
6       2024               SE              FT   
7       2024               SE              FT   
8       2024               MI              FT   
9       2024               MI              FT   

                         job_title  salary salary_currency  salary_in_usd  \
0                 Research Analyst  200000             USD         200000   
1                 Research Analyst  150000             USD         150000   
2        Machine Learning Engineer  220000             USD         220000   
3        Machine Learning Engineer  137500             USD         137500   
4                     Data Science  170000             USD         170000   
5                     Data Science  130000             USD         130000   
6    Business Intelligence Analyst  138000             USD         138000   
7    Business Intelligence Analyst   86000             USD          86000   
8  Business Intelligence Developer  105000             USD         105000   
9  Business Intelligence Developer   80000             USD          80000   

  employee_residence  remote_ratio company_location company_size  
0                 US             0               US            M  
1                 US             0               US            M  
2                 US             0               US            M  
3                 US             0               US            M  
4                 US             0               US            M  
5                 US             0               US            M  
6                 US             0               US            M  
7                 US             0               US            M  
8                 US           100               US            M  
9                 US           100               US            M

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12688 entries, 0 to 12687
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           12688 non-null  int64 
 1   experience_level    12688 non-null  object
 2   employment_type     12688 non-null  object
 3   job_title           12688 non-null  object
 4   salary              12688 non-null  int64 
 5   salary_currency     12688 non-null  object
 6   salary_in_usd       12688 non-null  int64 
 7   employee_residence  12688 non-null  object
 8   remote_ratio        12688 non-null  int64 
 9   company_location    12688 non-null  object
 10  company_size        12688 non-null  object
dtypes: int64(4), object(7)
memory usage: 1.1+ MB


In [4]:
df.drop_duplicates().info()

<class 'pandas.core.frame.DataFrame'>
Index: 7751 entries, 0 to 12687
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           7751 non-null   int64 
 1   experience_level    7751 non-null   object
 2   employment_type     7751 non-null   object
 3   job_title           7751 non-null   object
 4   salary              7751 non-null   int64 
 5   salary_currency     7751 non-null   object
 6   salary_in_usd       7751 non-null   int64 
 7   employee_residence  7751 non-null   object
 8   remote_ratio        7751 non-null   int64 
 9   company_location    7751 non-null   object
 10  company_size        7751 non-null   object
dtypes: int64(4), object(7)
memory usage: 726.7+ KB


Dropping duplicates drops 4936 columns which leaves ~60% of the data intact. Considering the data is user collected, this could be from clicking the submit button multiple times, submitting salaries multiple times on purpose or perhaps entries that are not actually duplicates. All the duplicates will be dropped as there no justification for keeping them.

The datasource has 11 columns and 7751 rows after removing duplicates.
There is no missing data in the dataset.

`work_year`, `salary`, `salary_in_usd`, `remote_ratio` are int64

All other columns are strings

In [5]:
def sorted_list_col(column):
    print('{entries} unique entries in the {column} column \n'.format(entries=len(df[column].unique()), column=column))
    print(np.sort(df[column].unique()))


In [6]:
sorted_list_col('job_title')

139 unique entries in the job_title column 

['AI Architect' 'AI Developer' 'AI Engineer' 'AI Product Manager'
 'AI Programmer' 'AI Research Engineer' 'AI Scientist'
 'AWS Data Architect' 'Admin & Data Analyst' 'Analytics Engineer'
 'Analytics Engineering Manager' 'Applied Data Scientist'
 'Applied Machine Learning Engineer' 'Applied Machine Learning Scientist'
 'Applied Scientist' 'Autonomous Vehicle Technician' 'Azure Data Engineer'
 'BI Analyst' 'BI Data Analyst' 'BI Data Engineer' 'BI Developer'
 'Big Data Architect' 'Big Data Engineer' 'Business Data Analyst'
 'Business Intelligence' 'Business Intelligence Analyst'
 'Business Intelligence Data Analyst' 'Business Intelligence Developer'
 'Business Intelligence Engineer' 'Business Intelligence Lead'
 'Business Intelligence Manager' 'Business Intelligence Specialist'
 'Cloud Data Architect' 'Cloud Data Engineer' 'Cloud Database Engineer'
 'Compliance Data Analyst' 'Computer Vision Engineer'
 'Computer Vision Software Engineer' 'Consu

I am noticing that there is some overlap in the job titles. Here are a few examples:
1. BI Analyst, BI Data Analyst, Business Intelligence Analyst and Business Intelligence Data Analyst
2. Data Modeler, Data Modeller
3. Computer Vision Engineer, Computer Vision Software Engineer
4. MLOps Engineer, Machine Learning Operations Engineer
5. Product Data Analyst vs Sales Data Analyst?

Consolidating the job titles will be an effective way to improve the model performance. A good idea might be to build a dictionary with each title and all the sub titles that exist in this dataset that way a mapping function can be used to quickly create a new column with consolidated job descriptions. This will be explored fully in the EDA stage of the project.

Another idea might be to identify keywords in the job titles e.g. Engineer vs Analyst vs Scientist or Principal vs Staff vs Lead to help delineate levels and roles and understand ther impact on the salary.

In [7]:
sorted_list_col('experience_level')

4 unique entries in the experience_level column 

['EN' 'EX' 'MI' 'SE']


This is a straightforward column with only 4 entries. During feature engineering, these can be assigned to integers ordered by increasing seniority.

EN- entry level
EX- executive level
MI- mid level
SE- senior level

In [8]:
sorted_list_col('employment_type')

4 unique entries in the employment_type column 

['CT' 'FL' 'FT' 'PT']


Like the `experience_level` column, there are only 4 possibilities. Unlike the previous example, these should probably be listed as categorical during feature engineering as there is not necessary expected that salary will correlate directly with employment type

CT- contract
FL- freelance
FT- full time
PT- part time

In [9]:
sorted_list_col('employee_residence')

87 unique entries in the employee_residence column 

['AD' 'AE' 'AM' 'AR' 'AS' 'AT' 'AU' 'BA' 'BE' 'BG' 'BO' 'BR' 'CA' 'CF'
 'CH' 'CL' 'CN' 'CO' 'CR' 'CY' 'CZ' 'DE' 'DK' 'DO' 'DZ' 'EC' 'EE' 'EG'
 'ES' 'FI' 'FR' 'GB' 'GE' 'GH' 'GR' 'HK' 'HN' 'HR' 'HU' 'ID' 'IE' 'IL'
 'IN' 'IQ' 'IR' 'IT' 'JE' 'JP' 'KE' 'KR' 'KW' 'LB' 'LT' 'LU' 'LV' 'MD'
 'MT' 'MU' 'MX' 'MY' 'NG' 'NL' 'NO' 'NZ' 'PE' 'PH' 'PK' 'PL' 'PR' 'PT'
 'QA' 'RO' 'RS' 'RU' 'SA' 'SE' 'SG' 'SI' 'TH' 'TN' 'TR' 'UA' 'UG' 'US'
 'UZ' 'VN' 'ZA']


The data is collected from 87 unique countries. This offers a broad selection of countries however it would be important to see the distribution of data by country. I hypothesize that the country will substantially affect the salary of data professionals.

In [10]:
sorted_list_col('salary_currency')

23 unique entries in the salary_currency column 

['AUD' 'BRL' 'CAD' 'CHF' 'CLP' 'DKK' 'EUR' 'GBP' 'HKD' 'HUF' 'ILS' 'INR'
 'JPY' 'MXN' 'NOK' 'NZD' 'PHP' 'PLN' 'SGD' 'THB' 'TRY' 'USD' 'ZAR']


On it's own, the currency column would simply serve the purpose of converting the salary to USD using the exchange rate since the salaries are provided in local currency and will be dropped.

In [11]:
sorted_list_col('company_size')

3 unique entries in the company_size column 

['L' 'M' 'S']


This column has three possible values L-large, M-medium, S-small. No transformations are needed for this column.

In [12]:
sorted_list_col('company_location')

76 unique entries in the company_location column 

['AD' 'AE' 'AM' 'AR' 'AS' 'AT' 'AU' 'BA' 'BE' 'BR' 'BS' 'CA' 'CF' 'CH'
 'CL' 'CN' 'CO' 'CZ' 'DE' 'DK' 'DZ' 'EC' 'EE' 'EG' 'ES' 'FI' 'FR' 'GB'
 'GH' 'GI' 'GR' 'HK' 'HN' 'HR' 'HU' 'ID' 'IE' 'IL' 'IN' 'IQ' 'IR' 'IT'
 'JP' 'KE' 'KR' 'LB' 'LT' 'LU' 'LV' 'MD' 'MT' 'MU' 'MX' 'MY' 'NG' 'NL'
 'NO' 'NZ' 'PH' 'PK' 'PL' 'PR' 'PT' 'QA' 'RO' 'RU' 'SA' 'SE' 'SG' 'SI'
 'TH' 'TR' 'UA' 'US' 'VN' 'ZA']


This column consists of 76 unique countries, which is fewer than the number of coutries where the employees reside. A possible new feature would be whether the employee is based in the same location as the country's main office.

In [13]:
sorted_list_col('remote_ratio')

3 unique entries in the remote_ratio column 

[  0  50 100]


This column, notably only has three possible values which refer to 0-onsite, 50-hybrid, 100-remote. I will leave this column as is


In [14]:
df.drop('remote_ratio',axis=1).describe()

work_year        salary  salary_in_usd
count  12688.000000  1.268800e+04   12688.000000
mean    2022.993301  1.674197e+05  149785.706967
std        0.654000  3.834499e+05   67922.547445
min     2020.000000  1.400000e+04   15000.000000
25%     2023.000000  1.040000e+05  103762.250000
50%     2023.000000  1.422000e+05  141525.000000
75%     2023.000000  1.880250e+05  185900.000000
max     2024.000000  3.040000e+07  750000.000000

Regarding the numeric columns.

`work_year` The work year indicates that this dataset has collected data between 2020 and 2024

`salary` The data has a very wide range from 14,000 to 30,000,000. However these numbers are useless on their own as they do not consider purchasing power. Instead of focusing only on the salary in USD, it may be interesting to do a conversion with the data using the purchasing power for each country and year to get a better idea of the salary by country. Perhaps the purchase power parity would be useful feature for the model that could be merged with the existing dataframe.

`salary_in_usd` The range on this data column is much smaller than the raw salary column. The data ranges from 15,000 USD to 750,000 USD. With a median of 142,000 the data appears skewed however that may be due to an uneven distribution of seniority levels and countries in the dataset.


## Conclusions:
The dataset has no missing values however about 40% of the data is duplicated and the duplicates will need to be dropped.

The job title column has overlapping job titles and consolidation should improve the results from the model.



No other columns seem to have any unusual data in them.